tst/unit/test_state_descriptor.cpp

```diff
@@ -278,15 +278,123 @@ TEST_CASE("Test reqendency resolution in StateDescriptor", "[StateDescriptor]")
+
+        AND_THEN("We can't add the same sparse ID twice") { REQUIRE_THROWS(pool.Add(2)); }
+      }
+
+      AND_THEN("We can't add InvalidSparseID") {
+        REQUIRE_THROWS(pool.Add(parthenon::InvalidSparseID));
+      }
+    }
+
+    THEN("Trying to create a SparsePool with dense metadata throws an error") {
+      REQUIRE_THROWS(SparsePool("dense", dense));
+    }
+  }
+
+  GIVEN("An empty state descriptor") {
+    auto pkg = std::make_shared<StateDescriptor>("pkg");
+    Metadata meta_sparse({Metadata::Sparse});
+
+    THEN("We can add sparse pools in different ways") {
+      SparsePool pool1("pool1", meta_sparse);
+      pool1.Add(0);
+      pool1.Add(55);
+      REQUIRE(pkg->AddSparsePool(pool1));
+
+      REQUIRE(pkg->AddSparsePool("pool2", meta_sparse, std::vector<int>{1, 55, 100}));
+
+      std::vector<std::vector<int>> shapes;
+      shapes.push_back({3});
+      shapes.push_back({2, 2});
+      REQUIRE(pkg->AddSparsePool(
+          "pool3", meta_sparse, std::vector<int>{0, 100}, shapes,
+          std::vector<MetadataFlag>{Metadata::Vector, Metadata::Tensor}));
+
+      REQUIRE(pkg->FieldPresent("pool1", 0));
+      REQUIRE(pkg->FieldPresent("pool1", 55));
+      REQUIRE(pkg->FieldPresent("pool2", 1));
+      REQUIRE(pkg->FieldPresent("pool2", 55));
+      REQUIRE(pkg->FieldPresent("pool2", 100));
+      REQUIRE(pkg->FieldPresent("pool3", 0));
+      REQUIRE(pkg->FieldPresent("pool3", 100));
+
+      AND_THEN("We can't add a SparsePool with wrong number of Vector/Tensor flags") {
+        REQUIRE_THROWS(pkg->AddSparsePool(
+            "pool3", meta_sparse, std::vector<int>{0, 100}, shapes,
+            std::vector<MetadataFlag>{Metadata::Vector, Metadata::Tensor,
+                                      Metadata::Tensor}));
+      }
+    }
+
+    THEN("We can't add fields/pools with the same names") {
+      REQUIRE(pkg->AddField("dense", Metadata()));
+      REQUIRE_FALSE(pkg->AddField("dense", Metadata()));
+      REQUIRE_FALSE(pkg->AddSparsePool("dense", meta_sparse, std::vector<int>{1, 2}));
+
+      REQUIRE(pkg->AddSparsePool("sparse", meta_sparse, std::vector<int>{1, 2}));
+      REQUIRE_FALSE(pkg->AddSparsePool("sparse", meta_sparse, std::vector<int>{4, 5}));
+      REQUIRE_FALSE(pkg->AddField("sparse", Metadata()));
+      REQUIRE_FALSE(pkg->AddField("sparse_1", Metadata()));
+      REQUIRE_FALSE(pkg->AddField("sparse_2", Metadata()));
+
+      // this is ok, but probably not a good idea
+      REQUIRE(pkg->AddField("sparse_3", Metadata()));
+
+      // check the other way around
+      REQUIRE(pkg->AddField("fake_sparse_27", Metadata()));
+      REQUIRE(pkg->AddSparsePool("fake_sparse", meta_sparse, std::vector<int>{10, 12}));
+      REQUIRE_FALSE(
+          pkg->AddSparsePool("fake_sparse", meta_sparse, std::vector<int>{13, 27, 9}));
+
+      REQUIRE(pkg->AddField("fake2_sparse_27", Metadata()));
+      REQUIRE_THROWS(
+          pkg->AddSparsePool("fake2_sparse", meta_sparse, std::vector<int>{13, 27, 9}));
+    }
```

Comment:
> Is there not an instance when we would also like to add multiple dense fields? 

I believe this is tested elsewhere in the file and @jlippuner didn't need to change it as the API is unchanged.

> why not have AddField and AddFields methods for both sparse and dense types

There's no reason we couldn't do this. But it's not used anywhere downstream. I'd suggest resisting adding new features until we need them.

> I feel like we are missing a chance to abstract away some of implementation details here. Dense and sparse data are both data objects. If we can prevent using concrete classes and methods that are specific to the data types I believe it will make things more adaptable in the future and less prone to break.

I'm not sure I understand. Are you suggesting a "dense field pool" object and adding things the same way for sparse and dense? That seems overly complicated to me.

Reply:
The abstraction is already there, in a sense. There are no sparse and dense variables or fields. There are just `CellVariable`s. The whole sparse pool stuff is essentially a convenient front end for adding several `CellVariable`s that are all related (share a base name).

We add multiple dense fields by calling by calling `AddField` multiple times, as before (and that was already tested in this test, which I didn't change). I guess you are suggesting to add a version of `AddField` that takes a vector of names and/or metadata. We could do that, but I agree with @Yurlungur that there isn't really a need for it now and if downstream codes want that later, we can add it.